Attempting to add debug logs for ENQUEUING an invalid object

NHDaly opened the pull request:
We were seeing GC Corruption when _popping_ an invalid object, but we don't know how we got this invalid pointer. (And we can't repro on linux so can't use RR.)

This is an attempt to check for the object's validity _before enqueuing_ so that we can hopefully give a more useful error message (which object's pointer was corrupted).

But for some reason it's currently complaining about the very first push....

NHDaly committed:
Attempting to add debug logs for ENQUEUING an invalid object

We were seeing GC Corruption when _popping_ an invalid object, but we
don't know how we got this invalid pointer. (And we can't repro on linux
so can't use RR.)

This is an attempt to check for the object's validity _before enqueuing_
so that we can hopefully give a more useful error message (which
object's pointer was corrupted).

But for some reason it's currently complaining about the very first push....

NHDaly commented on src/gc.c:
```diff
@@ -1771,11 +1771,41 @@ STATIC_INLINE uintptr_t gc_read_stack(void *_addr, uintptr_t offset,
+    // jl_(vt);
+
+    //if (__unlikely(!jl_is_datatype(vt) || vt->smalltag)) {
+    if (__unlikely(!jl_is_datatype(vt))) {
```
CC: @d-netto and @vchuravy: I'm not sure why, but this is always triggering for me...

vchuravy replied:
Is this on top of master? If so #49556 bt @vtjnash has made this check more complicated.

NHDaly replied:
Yeah, it was on top of master.
Thanks.

I saw the assertion check was changed to `(__unlikely(!jl_is_datatype(vt) || vt->smalltag))` but removing the rhs of the `||` should make it trigger _less often_, not more often -- that's why i was confused.

Is there more that has to change for this?

NHDaly replied:
So is the idea something like this?:

This check here:
https://github.com/JuliaLang/julia/blob/31b1374ff632a5b767d1d6d873eb02d5b5288c0f/src/gc.c#L2569-L2573
was only happening as an _else block,_ after all these other possible cases. So it's not required that everything we push is a jl_value_t with a jl_datatype_t in its type tag - there are a bunch of other cases to check too.

Something like that?

NHDaly replied:
K, yeah, i think that's what's going on. Thanks for the pointer!
K, i'm just gonna checkout a commit from 6 days ago for now, then 😁

NHDaly replied:
Aha! It turned out the issue was the erroneous JL_NORETURN which i had accidentally copy/pasted from the function below. I've fixed that, and I also made the changes to account for Jameson's changes. Thank you!

NHDaly committed:
Fix bugs in assertion check

- Remove JL_NORETURN
- Account for changes in https://github.com/JuliaLang/julia/pull/49556

d-netto committed:
continue Nathan's work

d-netto commented:
Fixed some compiler warnings and inserted the `gc_assert_parent_validity` in the places where the parent validity should be checked when enqueueing an object.

On the example from https://github.com/JuliaLang/julia/pull/21590 (with `GC_ASSERT_PARENT_VALIDITY` defined):

```julia
a = Ref(1)
ptr = pointer_from_objref(a) - 8
unsafe_store!(Ptr{Int}(ptr), ptr)
GC.gc()
```
```
GC error (probable corruption)
Allocations: 15177 (Pool: 15157; Big: 20); GC: 0
Parent 0x11809dba0
of type:
Core.Binding
While marking child at 0x1180f4350
of type:
Core.ReturnNode(val=SSAValue(1))
```

The performance of keeping `GC_ASSERT_PARENT_VALIDITY` enabled needs to be further assessed in order to justify the decision of keeping under a debug flag or enabling it by default.

NHDaly commented on src/gc.c:
```diff
@@ -2557,10 +2547,8 @@ FORCE_INLINE void gc_mark_outrefs(jl_ptls_t ptls, jl_gc_markqueue_t *mq, void *_
-                        assert_validity_of_new_obj(parent, new_obj);
+                    else
                         gc_ptr_queue_push(mq, new_obj);
-                    }
```
Why were you able to delete all of these? Is it because you moved the assertion to happen before this whole function is called?

d-netto replied:
The `gc_assert_parent_validity` in `gc_mark_obj{8, 16, 32}` should already cover these cases.

NHDaly reviewed (commented):
Awesome. Thanks again for picking this up and carrying it forward, @d-netto!! :)

The hardest part to me will be ensuring that we don't ever miss a spot and push an object without including this assertion there. (Whereas before, when it was on the _pop_ side, it was easier to ensure it always happens.) But I think it should be pretty doable.

Thanks!
